src/mumble/AudioOutput.cpp

```diff
@@ -168,35 +183,69 @@ void AudioOutput::addFrameToBuffer(ClientUser *sender, const Mumble::Protocol::A
 
 		qrwlOutputs.lockForWrite();
 
-		aop = new AudioOutputSpeech(sender, iMixerFreq, audioData.usedCodec, iBufferSize);
-		qmOutputs.replace(sender, aop);
+		speech = new AudioOutputSpeech(sender, iMixerFreq, audioData.usedCodec, iBufferSize);
+		qmOutputs.replace(sender, speech);
 	}
 
-	aop->addFrameToBuffer(audioData);
+	speech->addFrameToBuffer(audioData);
 
 	qrwlOutputs.unlock();
 }
 
-void AudioOutput::removeBuffer(const ClientUser *user) {
-	removeBuffer(qmOutputs.value(user));
-}
-
-void AudioOutput::removeBuffer(AudioOutputUser *aop) {
+void AudioOutput::handleInvalidatedBuffer(AudioOutputBuffer *buffer) {
 	QWriteLocker locker(&qrwlOutputs);
-	QMultiHash< const ClientUser *, AudioOutputUser * >::iterator i;
+	QMultiHash< const ClientUser *, AudioOutputBuffer * >::iterator i;
 	for (i = qmOutputs.begin(); i != qmOutputs.end(); ++i) {
-		if (i.value() == aop) {
+		if (i.value() == buffer) {
 			qmOutputs.erase(i);
-			delete aop;
+			delete buffer;
+			break;
+		}
+	}
+}
+
+void AudioOutput::handlePositionedBuffer(AudioOutputBuffer *buffer, float x, float y, float z) {
+	QReadLocker locker(&qrwlOutputs);
+	QMultiHash< const ClientUser *, AudioOutputBuffer * >::iterator i;
+	for (auto iter = qmOutputs.begin(); iter != qmOutputs.end(); ++iter) {
+		if (*iter == buffer) {
+			buffer->fPos[0] = x;
+			buffer->fPos[1] = y;
+			buffer->fPos[2] = z;
 			break;
 		}
 	}
 }
 
-AudioOutputSample *AudioOutput::playSample(const QString &filename, float volume, bool loop) {
+void AudioOutput::setBufferPosition(const AudioOutputToken &token, float x, float y, float z) {
+	if (token.m_buffer == nullptr) {
+		return;
+	}
+
+	emit bufferPositionChanged(token.m_buffer, x, y, z);
+}
+
+void AudioOutput::removeBuffer(AudioOutputBuffer *buffer) {
+	if (buffer == nullptr) {
+		return;
+	}
+
+	emit bufferInvalidated(buffer);
+}
+
+void AudioOutput::removeUser(const ClientUser *user) {
+	removeBuffer(qmOutputs.value(user));
+}
+
+void AudioOutput::removeToken(AudioOutputToken &token) {
+	removeBuffer(token.m_buffer);
+	token.m_buffer = nullptr;
```

Comment:
Why do we perform extra work in this function compared to the other ones?

Reply:
Well this is the only API for the user to remove/destroy the buffer. I think it makes sense to automatically set the pointer to null. Otherwise the user of the API would have to change their value to a nullptr-Token by hand everytime removeToken is called.

Reply:
Well, such a behavior would be the same as when deleting a pointer. The deletion itself only deletes the pointer but doesn't change the value of the pointer you provided it with. If you want that pointer to be `nullptr` afterwards, you have to do that yourself.
However, I suppose that if we require this nulling-behavior anyway, we can also let the function take care of it. In this case though, it might be better to use
```suggestion
void AudioOutput::removeToken(AudioOutputToken &token) {
	removeBuffer(token.m_buffer);
	token = {};
```
or
```suggestion
void AudioOutput::removeToken(AudioOutputToken &token) {
	removeBuffer(token.m_buffer);
	token.reset();
```